doc/comitup.8
.\" Automatically generated by Pandoc 1.17.2
.\"
.TH "comitup" "8" "April 2018" "" ""
.hy
.SH NAME
.PP
comitup \-\- Manage wifi connections on headless, unconnected systems
.SS SYNOPSIS
.PP
\f[C]comitup\ [options]\f[]
.SS DESCRIPTION
.PP
The \f[B]comitup\f[] service provides a means to establish a connection
between a computer and a wifi access point, in the case where wifi is
the only means available to access the computer.
.PP
On startup, the service will attempt to connect to wifi using
established networkmanager(8) connections.
If this is not successful, \f[B]comitup\f[] will establish a wifi
hotspot with the name \f[I]comitup\-<nnnn>\f[], where <nnnn> is a
persistent 4\-digit number.
.PP
While the hotspot is active, a comitup\-web(8) service is available to
manage connecting to an access point.
.PP
If two wifi interfaces are available, the hotspot will remain active on
the first interface, and the internet connection will be made on the
second.
Otherwise, the hotspot will be replaced with the internet connection.
.PP
In all states, avahi\-daemon(8) is used to publish the mdns host name
\f[I]comitup\-<nnnn>.local\f[], making the web service accessible as
e.g.
\f[I]http://comitup\-1234.local\f[], for systems supporting Zeroconf.
For other systems, a \f[I]comitup\f[] Workstation entry is published
which is visible to Zeroconf browsing applications, allowing the IP
address to be manually determined.
The web service address is likely \f[I]http://10.42.0.1\f[].
.PP
\f[B]comitup\f[] logs to \f[I]/var/log/comitup.log\f[].
.SS Options
.IP \[bu] 2
\f[I]\-h\f[], \f[I]\-\-help\f[] \- Print help and exit
.IP \[bu] 2
\f[I]\-c\f[], \f[I]\-\-check\f[] \- Check the wifi device configuration
and exit
.SS D\-Bus Interface
.PP
\f[B]Comitup\f[] provides a D\-Bus object which claims the name
\f[I]com.github.davesteele.comitup\f[] on the path
\f[I]/com/github/davesteele/comitup\f[], supporting the interface
\f[I]com.github.davesteele.comitup\f[].
The interface includes the following methods.
.IP \[bu] 2
\f[I]get_info()\f[]
.RS 2
.PP
Input: None
.PP
Output: \f[I]DICT_ENTRY\f[]
.PP
Return information about the current \f[B]Comitup\f[] service.
The keys are as follows:
.IP \[bu] 2
\f[I]version\f[] \- The package version.
.IP \[bu] 2
\f[I]basename\f[] \- The currently configured basename (default
\f[B]comitup\f[]).
.IP \[bu] 2
\f[I]id\f[] \- The unique random numeric id associated with the service
instance.
.IP \[bu] 2
\f[I]hostnames\f[] \- A list of host names that are published for the
service IP address.
.IP \[bu] 2
\f[I]imode\f[] \- The current interface mode for comitup.
This returns the string \[aq]single\[aq] or \[aq]router\[aq].
In \[aq]single\[aq] mode, the hotspot is terminated when
\f[B]CONNECTED\f[].
In \[aq]router\[aq] mode, the hotspot is retained, the upstream
connection is made with the other wifi device, and traffic is routed
between them.
The web service is terminated when \f[B]CONNECTED\f[].
.RE
.IP \[bu] 2
\f[I]access_points()\f[]
.RS 2
.PP
Input: None
.PP
Output: Array of \f[I]DICT_ENTRY\f[]
.PP
Return a list of visible access points.
This is represented as an array of D\-Bus \f[I]DICT_ENTRY\f[].
Each \f[I]DICT_ENTRY\f[] contains strings associated with the following
keys, \f[I]ssid\f[], \f[I]strength\f[] (0 to 100) and \f[I]security\f[]
(\f[I]encrypted\f[] or \f[I]unencrypted\f[]).
.RE
.IP \[bu] 2
\f[I]state()\f[]
.RS 2
.PP
Input: None
.PP
Output: \f[I]state\f[], \f[I]connection\f[]
.PP
This returns strings for the current \f[B]comitup\f[] state (either
\f[B]HOTSPOT\f[], \f[B]CONNECTING\f[], or \f[B]CONNECTED\f[]) and the
\f[I]ssid\f[] name for the current connection on the wifi device.
.RE
.IP \[bu] 2
\f[I]connect()\f[]
.RS 2
.PP
Input: \f[I]ssid\f[], \f[I]password\f[]
.PP
Output: None
.PP
Delete any existing connection for \f[I]ssid\f[], create a new
connection, and attempt to connect to it.
The password may be a zero length string if not needed.
.RE
.IP \[bu] 2
\f[I]delete_connection()\f[]
.RS 2
.PP
Input: \f[I]ssid\f[]
.PP
Output: None
.PP
Delete the connection for \f[I]ssid\f[].
The system will not be able to reconnect using this connection.
.RE
.SS COPYRIGHT
.PP
Comitup is Copyright (C) 2016\-2017 David Steele <steele\@debian.org>.
.SS SEE ALSO
.PP
comitup\-conf(5), comitup\-cli(1), comitup\-web(8)


comitup/comitup.py
#!/usr/bin/python3
# Copyright (c) 2017-2018 David Steele <dsteele@gmail.com>
#
# SPDX-License-Identifier: GPL-2+
# License-Filename: LICENSE


import os
import logging
from logging.handlers import TimedRotatingFileHandler
from comitup import persist
from comitup import config
import random
import argparse
import sys


from gi.repository.GLib import MainLoop
from dbus.mainloop.glib import DBusGMainLoop
DBusGMainLoop(set_as_default=True)

from comitup import statemgr     # noqa
from comitup import webmgr       # noqa
from comitup import iptmgr       # noqa
from comitup import wificheck    # noqa

PERSIST_PATH = "/var/lib/comitup/comitup.json"
CONF_PATH = "/etc/comitup.conf"
LOG_PATH = "/var/log/comitup.log"


def deflog():
    log = logging.getLogger('comitup')
    log.setLevel(logging.INFO)
    handler = TimedRotatingFileHandler(
                LOG_PATH,
                encoding='utf=8',
                when='D',
                interval=7,
                backupCount=8,
              )
    fmtr = logging.Formatter(
            "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
           )
    handler.setFormatter(fmtr)
    log.addHandler(handler)

    return log


def load_data():
    conf = config.Config(
                CONF_PATH,
                defaults={
                    'base_name': 'comitup',
                    'web_service': '',
                    'external_callback': '/usr/local/bin/comitup-callback',
                },
             )

    data = persist.persist(
                PERSIST_PATH,
                {'id': str(random.randrange(1000, 9999))},
           )

    return (conf, data)


def inst_name(conf, data):
    return conf.base_name + '-' + data.id


def parse_args():
    parser = argparse.ArgumentParser(description="")

    parser.add_argument('-c', '--check', action='store_true',
            help="Check the wifi devices and exit")

    args = parser.parse_args()

    return args


def main():
    if os.geteuid() != 0:
        exit("Comitup requires root privileges")

    args = parse_args()

    log = deflog()
    log.info("Starting comitup")

    (conf, data) = load_data()

    if args.check:
        if wificheck.run_checks():
            sys.exit(1)
        else:
            sys.exit(0)
    else:
        wificheck.run_checks(verbose=False)

    webmgr.init_webmgr(conf.web_service)
    iptmgr.init_iptmgr()

    statemgr.init_state_mgr(
                conf, data,
                [webmgr.state_callback, iptmgr.state_callback],
             )

    loop = MainLoop()
    loop.run()


if __name__ == '__main__':
    main()
